### PR TITLE
Add more debugging for why audio ring/ringback might not be playing

### PR DESCRIPTION
**Pull request by MadLittleMods**
Add more debugging for why audio might not be playing (check if muted, check for error events)

More debugging for https://github.com/vector-im/element-web/issues/20832 and https://github.com/vector-im/element-web/issues/23211

We now log certain error type events that are emitted from the `<audio>` elements such as `error`, `emptied`, `stalled`, `stalled`, `waiting` (see the [mdn docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/audio#events) for more info about these).

You can also optionally turn on some more event debugging by running this in the devtools console:

Turn on:
```js
mxSettingsStore.setValue('debug_legacy_call_handler', null, 'device', true);
```

Turn off:
```js
mxSettingsStore.setValue('debug_legacy_call_handler', null, 'device', false);
```

## Dev notes

 - https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement
 - https://reactjs.org/docs/events.html#media-events
 - https://webreflection.medium.com/dom-handleevent-a-cross-platform-standard-since-year-2000-5bf17287fd38

---

We can also try listening for some of the problem events.

```js
type MediaEventType = keyof HTMLMediaElementEventMap;
const MEDIA_ERROR_EVENT_TYPES: MediaEventType[] = [
    'error',
    'emptied',
    'stalled',
    'suspend',
    'waiting',
];

// The native event listener pattern
private handleEvent(e: Event) {
    const target = e.target as HTMLElement;
    const audioId = target?.id;

    if (MEDIA_ERROR_EVENT_TYPES.includes(e.type)) {
        logger.error(`LegacyCallHandler: encountered ${e.type} event with ${audioId}`, e);
    }
}


MEDIA_ERROR_EVENT_TYPES.forEach((errorEventType) => {
    audio.addEventListener(errorEventType, this);
});
```

## Checklist

* [x] Tests written for new code (and old code if feasible)
* [ ] Linter and other CI checks pass
* [ ] ~~Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))~~

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

### CODE REVIEW

File: src/LegacyCallHandler.tsx [hunk 1]
```diff
@@ -71,13 +71,51 @@ export const PROTOCOL_SIP_VIRTUAL = 'im.vector.protocol.sip_virtual';
...
     Ringback = 'ringbackAudio',
     CallEnd = 'callendAudio',
     Busy = 'busyAudio',
 }
 
+const debuglog = (...args: any[]) => {
+    if (SettingsStore.getValue("debug_legacy_call_handler")) {
+        logger.log.call(console, "LegacyCallHandler debuglog:", ...args);
+    }
+};
```
Comment (MadLittleMods) [inline src/LegacyCallHandler.tsx:117]: Copied from 

https://github.com/matrix-org/matrix-react-sdk/blob/5f6b1dda8df7fb2a6939c32da75b020efc24ea7c/src/components/structures/ScrollPanel.tsx#L38-L42

File: src/LegacyCallHandler.tsx [hunk 2]
```diff
@@ -176,13 +215,55 @@ export default class LegacyCallHandler extends EventEmitter {
...
+
+    public handleEvent(e: Event) {
+        const target = e.target as HTMLElement;
+        const audioId = target?.id;
+
+        if (MEDIA_ERROR_EVENT_TYPES.includes(e.type as MediaEventType)) {
+            logger.error(`LegacyCallHandler: encountered "${e.type}" event with <audio id="${audioId}">`, e);
+        } else if (MEDIA_EVENT_TYPES.includes(e.type as MediaEventType)) {
+            debuglog(`encountered "${e.type}" event with <audio id="${audioId}">`, e);
+        }
```
Comment (MadLittleMods) [inline src/LegacyCallHandler.tsx:266]: Need exception for SoncarCloud code coverage, https://sonarcloud.io/summary/new_code?id=matrix-react-sdk&pullRequest=9642

No point in adding tests to make sure we emit debug logs at this level.

Comment (t3chguy) [reply src/LegacyCallHandler.tsx:266]: For debug sections try adding istanbul magic comments

Comment (MadLittleMods) [reply src/LegacyCallHandler.tsx:268]: Great shout @t3chguy! Looks like that did it 🎉 ✅ https://github.com/matrix-org/matrix-react-sdk/pull/9642/checks?check_run_id=9789591092

```
/* istanbul ignore next */
/* istanbul ignore if */
/* istanbul ignore else */
/* istanbul ignore start */
/* istanbul ignore end */
```

(only added it around the debug log parts and a test for the extra mute logic branch)

File: src/LegacyCallHandler.tsx [hunk 3]
```diff
@@ -119,6 +158,7 @@ export default class LegacyCallHandler extends EventEmitter {
     // call with a different party to this one.
     private transferees = new Map<string, MatrixCall>(); // callId (target) -> call (transferee)
     private audioPromises = new Map<AudioID, Promise<void>>();
+    private audioElementsWithListeners = new Map<HTMLMediaElement, boolean>();
     private supportsPstnProtocol = null;
```
Comment (MadLittleMods) [inline src/LegacyCallHandler.tsx:162]: The `Typescript Strict Error Checker (--noImplicitAny)` CI job is failing against unrelated adjacent lints, https://github.com/matrix-org/matrix-react-sdk/actions/runs/3580627599/jobs/6022910920

Can we ignore?

**Review by gsouquet — APPROVED** [review]

File: src/LegacyCallHandler.tsx [hunk 4]
```diff
@@ -176,13 +216,56 @@ export default class LegacyCallHandler extends EventEmitter {
...
+    private removeEventListenersForAudioElement(audioElement: HTMLMediaElement) {
+        MEDIA_EVENT_TYPES.forEach((errorEventType) => {
+            audioElement.removeEventListener(errorEventType, this);
+        });
+    }
+
+    /* istanbul ignore next */
+    public handleEvent(e: Event) {
+        const target = e.target as HTMLElement;
+        const audioId = target?.id;
```
Comment (t3chguy) [inline src/LegacyCallHandler.tsx:262]: `handleEvent` doesn't seem like a debug function so shouldn't be ignored in coverage, maybe `debugEvent`?

Comment (MadLittleMods) [reply src/LegacyCallHandler.tsx:262]: It's only used for debug logging stuff though. I did have a inkling to add a comment there to make sure to remove/move the magic ignore comment if we start using it for other things though.

File: src/LegacyCallHandler.tsx (hunk 3 above)
Comment (MadLittleMods) [reply src/LegacyCallHandler.tsx:162]: Asked out-of-band in an [internal room](https://matrix.to/#/!fLeHeojWgBGJlLNdLC:matrix.org/$MRXinLRLZ0J0sxsRUm7jWAIUB2xqnyVGsE4WtKWZB00?via=matrix.org&via=element.io&via=vector.modular.im) and @t3chguy also agreed that these are unrelated ⏩

File: src/LegacyCallHandler.tsx [hunk 5]
```diff
@@ -176,13 +216,56 @@ export default class LegacyCallHandler extends EventEmitter {
...
             cli.removeListener(CallEventHandlerEvent.Incoming, this.onCallIncoming);
         }
+
+        // Remove event listeners for the <audio> elements
+        Array.from(this.audioElementsWithListeners.keys()).forEach((audioElement) => {
+            this.removeEventListenersForAudioElement(audioElement);
+        });
+    }
+
+    private addEventListenersForAudioElement(audioElement: HTMLMediaElement) {
```
Comment (MadLittleMods) [inline src/LegacyCallHandler.tsx:243]: @t3chguy spotted that these need return types (and for all the new functions here).

The linter will soon yell about this too.

File: src/LegacyCallHandler.tsx (hunk 4 above)
Comment (MadLittleMods) [reply src/LegacyCallHandler.tsx]: Added a comment 👍

**Review by t3chguy — APPROVED** [review]

**Comment by MadLittleMods** [comment]
Thanks for the review and merge @gsouquet and @t3chguy 🐘

**Pull request merged**